### DISCUSSION TITLE
Extraction into library / collaboration

### DISCUSSION THREAD

**Discussion by jansorg** [redhat-developer/lsp4ij#594] [General]
First of all, thank you for providing an open-source LSP client for JetBrains IDEs!
It's a welcome replacement for the various older attempts...

For my Swift Support plugin, I've written my own LSP client implementation (using Kotlin). I don't want to maintain my own solution unless I have to...
Even though your implementation seems suitable for me, your solution currently has a few shortcomings which prevents me from dropping my own implementation:
- Your solution can only be used with a required dependency on your 3rd-party lsp4ij plugin. Unfortunately, that's a no-go for my own plugin and also for commercial plugins I'm working on for clients.
- I'm not seeing a way to implement LSP-based unit or integration tests in a plugin based on your solution

My preferred solution would be a Java library, which can be linked as a dependency into my own plugin, based on your lsp4ij solution.
But that would require a restructuring of your code into a multi-project Gradle build to build a library and your lsp4ij plugin as before. Plugins using only the library would link to it and enhance their own plugin.xml with the needed changes.

So my question is:
Would you support the extraction of your core LSP client code into a library?

With your support, I would do most of the work.
My implementation also support call hierarchy, method hierarchy, etc. and I'd be willing to contribute to your code base.

I'd appreciate a response and your consideration. Thank you!

**Comment by angelozerr** [comment]
> First of all, thank you for providing an open-source LSP client for JetBrains IDEs!

Thanks!

> It's a welcome replacement for the various older attempts...
> 
> For my Swift Support plugin, I've written my own LSP client implementation (using Kotlin). I don't want to maintain my own solution unless I have to...
> Even though your implementation seems suitable for me, your solution currently has a few shortcomings which prevents me from dropping my own implementation:
> - Your solution can only be used with a required dependency on your 3rd-party lsp4ij plugin. Unfortunately, that's a no-go for my own plugin and also for commercial plugins I'm working on for clients.

The main reason that we did that is because LSP4IJ contribute to the UI with LSP console which is required feature when you work with language server.

> - I'm not seeing a way to implement LSP-based unit or integration tests in a plugin based on your solution
> 

We have several tests at https://github.com/redhat-developer/lsp4ij/tree/main/src/test/java/com/redhat/devtools/lsp4ij/features which uses a mock language server. We could improve this code to support a custom language server and we could move thoses classes from test to a testFramework to allow using this class to test any language servers.

> My preferred solution would be a Java library, which can be linked as a dependency into my own plugin, based on your lsp4ij solution.

I dont know if it is possible. @fbricon what do you think about that?

> But that would require a restructuring of your code into a multi-project Gradle build to build a library and your lsp4ij plugin as before. Plugins using only the library would link to it and enhance their own plugin.xml with the needed changes.

Yes it is my fear that it is not a trivial issue but if you want to try it, we are opened with any contribution.

> 
> So my question is:
> Would you support the extraction of your core LSP client code into a library?

As I said we are opened to do that but it is not our priority. We did this project to consume in our IJ Quarkus plugin and it is oir first priority. We have the luck to have a lot of projects which are based on LSP4IJ and we benefit from contribution like documentSymbols.

> 
> With your support, I would do most of the work.
> My implementation also support call hierarchy, method hierarchy, etc. and I'd be willing to contribute to your code base.

It would fantastic to have this lsp support contribution. I think this support should looks like to documentSymbol support.

> 
> I'd appreciate a response and your consideration. Thank you!

Hope we could work together.

**Reply by jansorg** [reply]
Thank you for your reply!
I'll think about it, but I probably will stick with my own LSP client implementation. It would be (too) much work for me to contribute and maintain the library extraction. And I'm favoring Kotlin and coroutines.

**Reply by angelozerr** [reply]
> Thank you for your reply!
> I'll think about it, but I probably will stick with my own LSP client implementation. It would be (too) much work for me to contribute and maintain the library extraction. And I'm favoring Kotlin and coroutines.
> 

That is very shame for LSP4IJ. To be honnest with you I don't see the problem with the lsp support without library.

Developping an lsp support is very hard and takes so many times. With the 0.8.0 we did a big improvement with performance when IJ is scanning files (starting and executing ls while scanning files slow down IJ and we dont start ls if scanning files is processing)

It is a sample of feature that you could benefit with LSP4IJ for instance.

I would really love working together.

**Reply by jansorg** [reply]
Sure, for many using LSP4IJ as a plugin is a great solution.
For my own software, I can't / don't want to depend on another plugin to allow it to be installed. It's too complicated for users and it's doesn't provide the API or binary stability I need. 

Java-only is fine, too. But in the long run Kotlin and coroutines will be favored by JetBrains, I think.

I'll think about the pros and cons and I'm sorry for the back and forth since my first post. I'll let you know in the case I change my mind. I already know that it's a lot of work :) 

I know that I and many others do appreciate the work your're doing with LSP4IJ! Thanks again!

**Reply by angelozerr** [reply]
> For my own software, I can't / don't want to depend on another plugin to allow it to be installed. It's too complicated for users and it's doesn't provide the API or binary stability I need.

Ok.

> Java-only is fine, too. But in the long run Kotlin and coroutines will be favored by JetBrains, I think.

I wonder if Jetbrains LSP support is written in Kotlin or Java?

> I'll think about the pros and cons and I'm sorry for the back and forth since my first post. I'll let you know in the case I change my mind. I already know that it's a lot of work :)

Hope you will change your mind :) I have started to support call hierarchy in my spare time and if it is working I will try to embed itin the 0.8.0 release planned for next Monday 11/25/2024

> I know that I and many others do appreciate the work your're doing with LSP4IJ! Thanks again!

Thanks! Hope again that we could work together.

**Comment by angelozerr** [comment]
Just one question why you dont use the jetbrains lsp support since your plugin is a paid product?

**Reply by jansorg** [reply]
Not all users of a paid plugin use a paid IDE. And, for example, users on Android Studio could never use such a plugin.
And the JetBrains solution is closed-source and can't be extended with new features. There's too much missing to use it for a complete language integration.

**Reply by angelozerr** [reply]
Ok I understand.

**Comment by angelozerr** [comment]
@jansorg another question, have you tried LSP4IJ with the Swift language server 'without developping an IJ plugin)with https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserDefinedLanguageServer.md

If no I suggest that you read https://idetools.dev/blog/lsp4ij-announcement/

If yes, what do you think about the LSP support compare to your LSP support?

**Comment by angelozerr** [comment]
@jansorg I have implemented call+type hierarchy which will be available in 08.0 (it should be released on monday). 

Please read  LSP [callHierachy](https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#call-hierarchy)  and [typeHierarchy](https://github.com/redhat-developer/lsp4ij/blob/main/docs/LSPSupport.md#type-hierarchy) support to activate with your language.

I wonder how you have implemented method hierachy? Is it a custom feature with Swift language server, since I don't see LSP specification about method hierarchy?

**Comment by angelozerr** [comment]
@jansorg I have created a SourceToolkit-LSP template (see https://github.com/redhat-developer/lsp4ij/blob/main/docs/user-defined-ls/sourcekit-lsp.md) . You will need to install https://github.com/redhat-developer/lsp4ij?tab=readme-ov-file#testing-nightly-builds (or wait for Monday when release will be done).

You can test now very quickly SourceToolkit-LSP by using this template. It seems it is working but unfortunately as soon as I open completion, the language server seems broken and after that no LSP features like hover, code action, highlight etc doesn't give some LSP response.

Do you know what is the problem? Perhaps the command start `sourcetoolkit-lsp` must have more configuration? Thanks for your help.

**Comment by angelozerr** [comment]
@jansorg please note @SCWells72 works also for a commercial product and wants to use LSP4IJ. He has started several contributions like https://github.com/redhat-developer/lsp4ij/pull/637

**Reply by jansorg** [reply]
Thank you for letting me know and for the update for sourcekit-lsp. 
AFAIK Scott's use-case is to add support for additional/minor languages. My use-case is first-class support for the only language supported. 
I don't want to tell users that another plugin is needed to make things works at all. 
Perhaps, I just don't want to throw away my implementation 😉 But I still favor a library-approach with Kotlin.

iirc the command is just sourcekit-lsp. I remember crashes if the text sync was not entirely correct. But I don't remember blocking or it not working at all.

**Reply by angelozerr** [reply]
Ok thanks for your info.

Good luck for your project. Hope you will change your mind in the future :)

**Comment by angelozerr** [comment]
It seems @jansorg has had the same problem than me with sourcekit lsp https://youtrack.jetbrains.com/issue/IJPL-15762/LSP-API-apple-sourcekit-lsp-code-completion-doesnt-work-false-error-No-response-from-the-server-in-10-seconds

**Comment by angelozerr** [comment]
@FalsePattern I noticed that you have published LSP4IJ in maven central at https://mvnrepository.com/artifact/com.redhat.devtools.intellij/lsp4ij 

Why are you doing that? Is it the same reason than @jansorg to control the LSP4IJ version and not be linked to the LSP4IJ published in marketplace?
